File: apps/lib/services/location_service.dart
import 'dart:math';
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';

/// Privacy-first location service
/// - All distance calculations are done on-device
/// - Location is never sent to any server
/// - User must explicitly opt-in to location access
class LocationService extends ChangeNotifier {
  Position? _currentPosition;
  String? _currentCounty;
  bool _isLoading = false;
  String? _error;
  bool _hasPermission = false;

  // Getters
  Position? get currentPosition => _currentPosition;
  String? get currentCounty => _currentCounty;
  bool get isLoading => _isLoading;
  String? get error => _error;
  bool get hasPermission => _hasPermission;
  bool get hasLocation => _currentPosition != null;

  // Bay Area county centers for on-device county detection
  static const Map<String, Map<String, double>> _countyCoordinates = {
    'Alameda County': {'lat': 37.6017, 'lng': -121.7195},
    'Contra Costa County': {'lat': 37.9193, 'lng': -121.9277},
    'Marin County': {'lat': 38.0834, 'lng': -122.7633},
    'Napa County': {'lat': 38.5025, 'lng': -122.2654},
    'San Francisco': {'lat': 37.7749, 'lng': -122.4194},
    'San Mateo County': {'lat': 37.4969, 'lng': -122.3331},
    'Santa Clara County': {'lat': 37.3541, 'lng': -121.9552},
    'Solano County': {'lat': 38.2721, 'lng': -121.9399},
    'Sonoma County': {'lat': 38.5780, 'lng': -122.9888},
  };

  /// Check if location services are available
  Future<bool> checkPermission() async {
    try {
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        _error = 'Location services are disabled';
        return false;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        _hasPermission = false;
        return false;
      }

      if (permission == LocationPermission.deniedForever) {
        _error = 'Location permission permanently denied';
        _hasPermission = false;
        return false;
      }

      _hasPermission = true;
      return true;
    } catch (e) {
      _error = e.toString();
      return false;
    }
  }

  /// Request location permission
  Future<bool> requestPermission() async {
    try {
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        _error = 'Please enable location services';
        notifyListeners();
        return false;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
        if (permission == LocationPermission.denied) {
          _error = 'Location permission denied';
          _hasPermission = false;
          notifyListeners();
          return false;
        }
      }

      if (permission == LocationPermission.deniedForever) {
        _error = 'Location permission permanently denied. Please enable in Settings.';
        _hasPermission = false;
        notifyListeners();
        return false;
      }

      _hasPermission = true;
      _error = null;
      notifyListeners();
      return true;
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      return false;
    }
  }

  /// Get current location (on-device only)
  Future<bool> getCurrentLocation() async {
    if (_isLoading) return false;

    _isLoading = true;
    _error = null;
    notifyListeners();

    try {
      // Request permission if not granted
      if (!_hasPermission) {
        final granted = await requestPermission();
        if (!granted) {
          _isLoading = false;
          notifyListeners();
          return false;
        }
      }

      // Get position (stays on device!)
      _currentPosition = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.low, // Don't need precise location
        timeLimit: const Duration(seconds: 10),
      );

      // Determine county (calculated on-device)
      _currentCounty = _findNearestCounty(
        _currentPosition!.latitude,
        _currentPosition!.longitude,
      );

      _isLoading = false;
      notifyListeners();
      return true;
    } catch (e) {
      _error = 'Could not get location: ${e.toString()}';
      _isLoading = false;
      notifyListeners();
      return false;
    }
  }

  /// Clear current location
  void clearLocation() {
    _currentPosition = null;
    _currentCounty = null;
    _error = null;
    notifyListeners();
  }

  /// Find nearest county to given coordinates (on-device calculation)
  String _findNearestCounty(double lat, double lng) {
    String nearestCounty = 'Bay Area';
    double minDistance = double.infinity;

    for (final entry in _countyCoordinates.entries) {
      final distance = calculateDistance(
        lat, lng,
        entry.value['lat']!, entry.value['lng']!,
      );
      if (distance < minDistance) {
        minDistance = distance;
        nearestCounty = entry.key;
      }
    }

    return nearestCounty;
  }

  /// Calculate distance between two points using Haversine formula (on-device)
  /// Returns distance in miles
  static double calculateDistance(
    double lat1, double lng1,
    double lat2, double lng2,
  ) {
    const double earthRadiusMiles = 3959;

    final dLat = _toRadians(lat2 - lat1);
    final dLng = _toRadians(lng2 - lng1);

    final a = sin(dLat / 2) * sin(dLat / 2) +
        cos(_toRadians(lat1)) * cos(_toRadians(lat2)) *
        sin(dLng / 2) * sin(dLng / 2);

    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return earthRadiusMiles * c;
  }

  static double _toRadians(double degrees) {
    return degrees * pi / 180;
  }

  /// Format distance for display
  static String formatDistance(double miles) {
    if (miles < 0.1) {
      return '${(miles * 5280).round()} ft';
    } else if (miles < 10) {
      return '${miles.toStringAsFixed(1)} mi';
    } else {
      return '${miles.round()} mi';
    }
  }

  /// Calculate distance to a program location (on-device)
  double? getDistanceToProgram(double? programLat, double? programLng) {
    if (_currentPosition == null || programLat == null || programLng == null) {
      return null;
    }

    return calculateDistance(
      _currentPosition!.latitude,
      _currentPosition!.longitude,
      programLat,
      programLng,
    );
  }
}

File: apps/visionos/BayNavigator/Services/LocationService.swift
import Foundation
import CoreLocation

/// Privacy-first location service for visionOS
/// - All distance calculations are done on-device
/// - Location is never sent to any server
/// - User must explicitly opt-in to location access
@Observable
final class LocationService: NSObject {
    // MARK: - Published Properties
    private(set) var currentLocation: CLLocation?
    private(set) var currentCounty: String?
    private(set) var isLoading = false
    private(set) var error: String?
    private(set) var authorizationStatus: CLAuthorizationStatus = .notDetermined

    var hasLocation: Bool { currentLocation != nil }
    var hasPermission: Bool {
        authorizationStatus == .authorizedWhenInUse || authorizationStatus == .authorizedAlways
    }

    // MARK: - Private
    private let locationManager = CLLocationManager()

    // Bay Area county centers for on-device county detection
    private static let countyCoordinates: [String: CLLocationCoordinate2D] = [
        "Alameda County": CLLocationCoordinate2D(latitude: 37.6017, longitude: -121.7195),
        "Contra Costa County": CLLocationCoordinate2D(latitude: 37.9193, longitude: -121.9277),
        "Marin County": CLLocationCoordinate2D(latitude: 38.0834, longitude: -122.7633),
        "Napa County": CLLocationCoordinate2D(latitude: 38.5025, longitude: -122.2654),
        "San Francisco": CLLocationCoordinate2D(latitude: 37.7749, longitude: -122.4194),
        "San Mateo County": CLLocationCoordinate2D(latitude: 37.4969, longitude: -122.3331),
        "Santa Clara County": CLLocationCoordinate2D(latitude: 37.3541, longitude: -121.9552),
        "Solano County": CLLocationCoordinate2D(latitude: 38.2721, longitude: -121.9399),
        "Sonoma County": CLLocationCoordinate2D(latitude: 38.5780, longitude: -122.9888)
    ]

    // MARK: - Initialization
    override init() {
        super.init()
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyKilometer // Don't need precise
        authorizationStatus = locationManager.authorizationStatus
    }

    // MARK: - Public Methods

    /// Request location permission
    func requestPermission() {
        locationManager.requestWhenInUseAuthorization()
    }

    /// Get current location (on-device only - never sent to server)
    func getCurrentLocation() {
        guard !isLoading else { return }

        isLoading = true
        error = nil

        if authorizationStatus == .notDetermined {
            requestPermission()
            return
        }

        guard hasPermission else {
            error = "Location permission not granted"
            isLoading = false
            return
        }

        locationManager.requestLocation()
    }

    /// Clear current location
    func clearLocation() {
        currentLocation = nil
        currentCounty = nil
        error = nil
    }

    // MARK: - Distance Calculations (All On-Device)

    /// Calculate distance between two coordinates in miles
    static func calculateDistance(from: CLLocationCoordinate2D, to: CLLocationCoordinate2D) -> Double {
        let fromLocation = CLLocation(latitude: from.latitude, longitude: from.longitude)
        let toLocation = CLLocation(latitude: to.latitude, longitude: to.longitude)
        return fromLocation.distance(from: toLocation) / 1609.344 // meters to miles
    }

    /// Calculate distance from user to a program
    func distanceToProgram(latitude: Double?, longitude: Double?) -> Double? {
        guard let location = currentLocation,
              let lat = latitude,
              let lng = longitude else {
            return nil
        }

        let programCoord = CLLocationCoordinate2D(latitude: lat, longitude: lng)
        return Self.calculateDistance(from: location.coordinate, to: programCoord)
    }

    /// Format distance for display
    static func formatDistance(_ miles: Double) -> String {
        if miles < 0.1 {
            return "\(Int(miles * 5280)) ft"
        } else if miles < 10 {
            return String(format: "%.1f mi", miles)
        } else {
            return "\(Int(miles)) mi"
        }
    }

    // MARK: - Private Methods

    /// Find nearest county based on coordinates (on-device only)
    private func findNearestCounty(to coordinate: CLLocationCoordinate2D) -> String {
        var nearestCounty = "Bay Area"
        var minDistance = Double.infinity

        for (county, countyCoord) in Self.countyCoordinates {
            let distance = Self.calculateDistance(from: coordinate, to: countyCoord)
            if distance < minDistance {
                minDistance = distance
                nearestCounty = county
            }
        }

        return nearestCounty
    }
}

// MARK: - CLLocationManagerDelegate
extension LocationService: CLLocationManagerDelegate {
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }

        // Store location (stays on device!)
        currentLocation = location

        // Determine county (calculated on-device)
        currentCounty = findNearestCounty(to: location.coordinate)

        isLoading = false
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        self.error = "Could not get location: \(error.localizedDescription)"
        isLoading = false
    }

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        authorizationStatus = manager.authorizationStatus

        // If we were waiting for permission, try to get location now
        if isLoading && hasPermission {
            locationManager.requestLocation()
        } else if isLoading && authorizationStatus == .denied {
            error = "Location permission denied"
            isLoading = false
        }
    }
}
